"Add / Hide Tokens" screen

vladyslav-iosdev opened the pull request:
Closes #1747, closes #1739 

"Add / Hide Tokens" button in the Wallet tab
"Hide" button, and add hide token logic in the Wallet tab
Implement search on "Add / Hide Tokens" screen
Move add token logic to "Add / Hide Tokens" screen

hboon commented on AlphaWallet/Tokens/ViewModels/AddHideTokensViewModel.swift:
```diff
@@ -0,0 +1,225 @@
+        })
+    }
+    
+    private func getFilteredTokens(in tokens: [TokenObject], keyword: String?) -> [TokenObject] {
```
This looks quite similar to `TokensViewModel.getFilteredTokens()`. Can you extract a function out of them? Maybe stuff it as a static func in `TokensViewModel`. I'm normally not overly eager to extract common code because requirements might change and they need to be different, but the search keyword filtering logic should be in 1 place.

hboon commented on AlphaWallet/Tokens/Views/FungibleTokenViewCell.swift:
```diff
@@ -14,7 +15,10 @@ class FungibleTokenViewCell: UITableViewCell {
     private let blockChainTagLabel = UILabel()
     private let cellSeparators = (top: UIView(), bottom: UIView())
-
+    private var overlayViews: [UIView] {
```
Can you rename the property so it's intention-revealing?

hboon commented on AlphaWallet/Tokens/Views/TokensViewControllerTableViewSectionHeader.swift:
```diff
@@ -9,27 +9,67 @@ extension TokensViewController {
+                
                 NSLayoutConstraint.activate([
-                    filterView.anchorsConstraint(to: contentView, edgeInsets: .init(top: 0, left: 0, bottom: 7, right: 0)),
+                    filterView.widthAnchor.constraint(equalTo: self.widthAnchor),
```
Doesn't this change cause the horizontal and vertical position to be undefined? Unless it's anchored somewhere else?

vladyslav-iosdev replied:
No, it should work ok, for vertical and horizontal layouts

hboon replied:
Oh, you are right. I missed that it is added to a vertical `UIStackView`

hboon commented on AlphaWallet/Core/Initializers/MigrationInitializer.swift:
```diff
@@ -60,5 +60,37 @@ class MigrationInitializer: Initializer {
                 }
             }
         }
+        config.migrationBlock = { migration, oldSchemaVersion in
```
I understand this is a work in progress, but in the final commit, can you change the migration version to be just +1 (i.e. 6 instead of 8)?

Also, migration needs to take care of moving contracts in `HiddenContracts` over to `TokenObject` and marked as hidden

vladyslav-iosdev commented on AlphaWallet/Assets.xcassets/add_hide_tokens.imageset/Contents.json:
```diff
@@ -0,0 +1,23 @@
+{
```
1. You mean make it like this?
<img width="404" alt="Screenshot 2020-03-30 at 17 13 30" src="https://user-images.githubusercontent.com/55975226/77933639-afaaa800-72b7-11ea-8683-a7d7d5854cbd.png">


5. I don’t know, need debud this case.
7. When I tried to add this token I even don’t see it, it doesn’t appears in any tab, and I don’t see it in notifySubscribersOfUpdatedTokens
8. You mean add table view top inset or increase table view header height?
10. For now I didn’t find any ways to do that.

hboon replied:
1. No. Here's a screenshot of how it looks like in the Wallet tab. If you compare it to the screenshot above which I sent previously, focus on the sections I highlighted with the red "circle". There are 2 colors (grayish and white) in between the cells. Can you make the version in Add/Hide Tokens be like the one in the Wallet tab? 

<img src="https://user-images.githubusercontent.com/56189/77977984-7748af80-7333-11ea-9278-35c3a46f898b.png" width=230>

Bullet/Numbering is slightly off here, I'll stick to my original version in case I address the wrong question:

7. Did you tap on the + button in the Add/Hide Token screen (previously in Wallet tab) to bring up the screen with the title "Add Custom Token" to do this? This is the `NewTokenViewController` class.
8. I'm referring to this part. Note that this is 12.4, not 13.x. It looks ok in 13.x. Can you remove that additional margin?

<img width="230" src="https://user-images.githubusercontent.com/56189/77979885-6ea6a800-7338-11ea-8c57-1732710c786b.png">

9. If:

> For now I didn’t find any ways to do that.

you are referring to this:

> When we move a token from Hidden to Displayed, can we add it to the bottom of Displayed Tokens instead of to the top?

Can't this be done by setting the `sortIndex` accordingly?

vladyslav-iosdev replied:
Questions #9 I was wrong, the question was about 10. 
   i am not shure. sortIndex will not helo in this case.  As i understand correctrly, you want delete rows without confirmation, just by pressing "minus button", so i didn't find any ways to implement this, only with confirmation. If you know how to implement this, let me know.

7. yes i know how to add tokens, i tried to add token, and this feature works ok

hboon replied:
> 9. I was wrong, the question was about 10.
> i am not shure. sortIndex will not helo in this case. As i understand correctrly, you want delete rows without confirmation, just by pressing "minus button", so i didn't find any ways to implement this, only with confirmation. If you know how to implement this, let me know.

Ah. Don't worry about it then. Just leave it as that.

> 
> 1. yes i know how to add tokens, i tried to add token, and this feature works ok

Yes, so I was referring to adding the contract 0x8C0Edb69eBF038BA0c7a4873E40fC09725064c2e via this route here:

> Can you add this token on mainnet? 0x8C0Edb69eBF038BA0c7a4873E40fC09725064c2e. It should always have a balance of 1 I observe 2 issues with it. In the Add/Hide tokens screen, it is displayed as a balance of 0 (Wallet tab shows 1). And the "| Issuer: AlphaWallet" text is appear in Add/Hide tokens screen when it doesn't in the Wallet tab

vladyslav-iosdev commented on AlphaWallet/Tokens/ViewControllers/AddHideTokensViewController.swift:
```diff
@@ -0,0 +1,380 @@
+                            server: token.server,
+                            assetDefinitionStore: assetDefinitionStore,
+                            isVisible: isVisible,
+                            showDeleteConfirmation: false
```
Sorry, it's my fault. i forgot to remove that line.

vladyslav-iosdev commented on AlphaWallet/Core/Coordinators/FilterTokensCoordinator.swift:
```diff
@@ -0,0 +1,67 @@
+        return filteredTokens
+    }
+    
+    func sortDisplaydTokens(tokens: [TokenObject]) -> [TokenObject] {
```
you mean call it sortDisplayedTokens?

hboon replied:
Yes

vladyslav-iosdev commented on AlphaWallet/Tokens/ViewModels/AddHideTokensViewModel.swift:
```diff
@@ -0,0 +1,196 @@
+    private var displayedTokens: [TokenObject] = []
+    private var hiddenTokens: [TokenObject] = []
+    
+    private var searchText: String?
```
i will remove optional

vladyslav-iosdev commented on AlphaWallet/Tokens/ViewModels/AddHideTokensViewModel.swift:
```diff
@@ -0,0 +1,196 @@
+        }
+    }
+    
+    mutating func set(searchText text: String) {
```
i will change it

vladyslav-iosdev commented on AlphaWallet/Tokens/ViewModels/AddHideTokensViewModel.swift:
```diff
@@ -0,0 +1,196 @@
+        return tickers[token.server]?[token.contractAddress]
+    }
+    
+    private mutating func filter(tokens: [TokenObject], searchText: String?) {
```
ok, i will remove it

vladyslav-iosdev commented:
> 1. Can you make it so that the native crypto currency cards are at the top by default before all other token cardswith the database migration? Make them follow the order as in `ServersCoordinator.serversOrdered`. i.e If the user installs an app update which includes code in this PR, the Wallet tab should appear relatively unchanged (we don't need to sort all the tokens, but at least the native crypto cards shouldn't be moved away). They are free to move them around with the new Add / Hide Tokens screen after that.
> 
> <img alt="" width="200" src="https://user-images.githubusercontent.com/56189/78209733-6be3b880-74da-11ea-8172-c85cacab5586.png">
> 
> 1. In the Add / Hide Tokens screen, if you enter search mode (observed in iPhone Xs 12.4 simulator), noticed that first cell is clipped. Maybe the tableview content inset is wrong somehow? If you fix it in iOS 12.x, can you verify in iOS 13.x? These 2 iOS versions can have slightly different behavior here.
> 
> <img alt="" width="200" src="https://user-images.githubusercontent.com/56189/78209735-6f773f80-74da-11ea-94ec-29fc03f29b7e.png">
> 
> 1. Remember the space around "/" in "Add / Hide Tokens"
> 
> Looks good otherwise. Thanks!

1. I am not clearly undarstand how it should be. 
this new sort should be applied only on add/hide tokens screen, 
on hidden tokens section tokens should be sorted first, will be native tokens sorted by  ServersCoordinator.serversOrdered, and then rest of tokens? 
2. Fixed.
3. Fixed.

the rest of issues seems like fixed,  also one thing, you told about comments, what comments should i left, can you explain? and one more, about renaming variables and functions, next time can you provide name you want see there?

hboon commented:
@vladyslav-iosdev Oh sorry. I didn't realised I accidentally edited your comment above when I meant to reply. Here is my reply:

>> 1. Can you make it so that the native crypto currency cards are at the top by default before all other token cardswith the database migration? Make them follow the order as in `ServersCoordinator.serversOrdered`. i.e If the user installs an app update which includes code in this PR, the Wallet tab should appear relatively unchanged (we don't need to sort all the tokens, but at least the native crypto cards shouldn't be moved away). They are free to move them around with the new Add / Hide Tokens screen after that.

>1. I am not clearly undarstand how it should be. 
this new sort should be applied only on add/hide tokens screen, 
on hidden tokens section tokens should be sorted first, will be native tokens sorted by  ServersCoordinator.serversOrdered, and then rest of tokens? 

Notice that xDai and Ethereum are not at the top. Can you make Ethereum and xDai (i.e. the native crypto tokens) at the top instead immediately. Set the sortIndex if you have to in the database migration (following the order in `ServersCoordinator.serversOrdered`). But they should stay at the top until the user goes into the "Add / Hide Tokens" screen and change the order themselves. The other tokens doesn't matter. But the native crypto tokens are immportant and should be at the top

> you told about comments, what comments should i left, can you explain?

I'm not sure I did, unless you are referring to the commit message. If so, generally, squash the commits into atomic commits with commit messages that gives a short summary of the change and preferably a reason if necessary. For this PR, it looks like it's going to be just 1 commit, with the message ""Add / Hide Tokens" screen" (with no reason).

For another PR where it might be a simple change, but if there's a reason for making the change that is useful to document, we can say "Make submit button hit-test area bigger because users XXX". What is changed can usually be inferred from the source code diff given a 1-line summary in the commit message. The reason for the change is usually harder to infer.

> and one more, about renaming variables and functions, next time can you provide name you want see there?  

Ok

vladyslav-iosdev commented:
> @vladyslav-iosdev would you ping me in the PR if you want me to review the PR? Because I suppose you might be pushing the code as part of your workflow (to trigger Travis, to save them or for internal review, etc). I'll did it this time just in case you are blocked, but didn't want to comment unnecessarily on things you are planning to change/fix if you are in the midst of changing things.

@hboon i have question according sorting native tokens, 
as i understand required sorting already present in function notifySubscribersOfUpdatedTokens() right?
in this case we just need check if sorted tickets dont have sortIndex, thay should be left in order that was sorted in notifySubscribersOfUpdatedTokens()? if this is right i guess we dont need apply changes to migration function

hboon commented:
> @hboon i have question according sorting native tokens,
> as i understand required sorting already present in function notifySubscribersOfUpdatedTokens() right?

Yes. I have forgotten about the sorting done in `notifySubscribersOfUpdatedTokens()`. I supposed with this PR, the tokens are resort-ed again in `sortDisplaydTokens()`.

> in this case we just need check if sorted tickets dont have sortIndex, thay should be left in order that was sorted in notifySubscribersOfUpdatedTokens()? if this is right i guess we dont need apply changes to migration function

Let's say you have this before the PR, and in this order because of `notifySubscribersOfUpdatedTokens`:

```
Ethereum (native)
xDai (native)
TokenA
TokenB
```

After installing a build that includes this PR:

This still look like this: 

```
Ethereum
xDai
TokenA
TokenB
```

The user swaps TokenB and tokenA in the "Add / Hide Tokens" screen. Will it still display this?

```
Ethereum
xDai
TokenB
TokenA
```

And of course, they can reorder to be like this:

```
TokenB
TokenA
Ethereum
xDai
```

So as long as it can work like that, it doesn't matter if you set the sortIndex for the native tokens or not in the migration (I thought it's easier though). Up to you.

vladyslav-iosdev commented:
> > @hboon i have question according sorting native tokens,
> > as i understand required sorting already present in function notifySubscribersOfUpdatedTokens() right?
> 
> Yes. I have forgotten about the sorting done in `notifySubscribersOfUpdatedTokens()`. I supposed with this PR, the tokens are resort-ed again in `sortDisplaydTokens()`.
> 
> > in this case we just need check if sorted tickets dont have sortIndex, thay should be left in order that was sorted in notifySubscribersOfUpdatedTokens()? if this is right i guess we dont need apply changes to migration function
> 
> Let's say you have this before the PR, and in this order because of `notifySubscribersOfUpdatedTokens`:
> 
> ```
> Ethereum (native)
> xDai (native)
> TokenA
> TokenB
> ```
> 
> After installing a build that includes this PR:
> 
> This still look like this:
> 
> ```
> Ethereum
> xDai
> TokenA
> TokenB
> ```
> 
> The user swaps TokenB and tokenA in the "Add / Hide Tokens" screen. Will it still display this?
> 
> ```
> Ethereum
> xDai
> TokenB
> TokenA
> ```
> 
> And of course, they can reorder to be like this:
> 
> ```
> TokenB
> TokenA
> Ethereum
> xDai
> ```
> 
> So as long as it can work like that, it doesn't matter if you set the sortIndex for the native tokens or not in the migration (I thought it's easier though). Up to you.

@The user swaps TokenB and tokenA in the "Add / Hide Tokens" screen. Will it still display this?@

yes, after we do sort we take all displaying tokens and assign them sort Index according in array positioning, so in this case i can suppose that way i mentioned above will help reach wanted behavior

hboon commented:
@vladyslav-iosdev if the user swaps TokenA and TokenB only so that it looks like this in the Add/Hide Tokens screen:

```
Ethereum
xDai
TokenB
TokenA
```

Then it should look like this too in the Wallets tab.

If they drag the Ethereum and xDai token and so that it looks like this in the "Add / Hide Tokens" screen:

```
TokenB
TokenA
Ethereum
xDai
```

Then, the Wallet tab should follow too. The 2 screens should show the tokens in the same order.

vladyslav-iosdev commented:
> @vladyslav-iosdev looks good. Thanks! If there's nothing else missing, can you help to squash the commits into 1 with the message: ""Add / Hide Tokens" screen" and force push?

Yes np, can you explain what exactly should i do? Because i didn't work in way.

hboon commented:
> > @vladyslav-iosdev looks good. Thanks! If there's nothing else missing, can you help to squash the commits into 1 with the message: ""Add / Hide Tokens" screen" and force push?
> 
> Yes np, can you explain what exactly should i do? Because i didn't work in way.

Sure, that's fine. This is part of what PRs are about — exchange of knowledge.

But before I forget, we talked about combining the database migration versions into 1, instead of 6, 7, 8 (I think?), can you just keep them as `6`?

Squashing:

Run:

```
git log --pretty=oneline add_hide_tokes
```

shows:

```
75ae1d15669fc83f8d4473d22b5f2482057948da (add_hide_tokes) Minor refactor and bugs fix: - re...
5e54ab6ce3ca7fc92c84c903dff9f0fe5a5820a3 remove unused pod, fixed bugs, refactored code
0c06f1e241eeb6c23d61f676818859a3fa816b6e code refactor, bags fix
d03596eba69e470305228cb1d45bafa34bcb0c62 "Add / Hide Tokens" screen "Add / Hide Tok...
5c3f75c9ee0bb6da69ff38e4f7de4bc9b8b632fd Fixed src link of main image
```

Newest commit at the top, so there are 4 commits. We would like to combine them into 1.

Note the commit 1 older than the one you don't want to change: 5c3f75c9ee0bb6da69ff38e4f7de4bc9b8b632fd

So do:


```
git rebase -i 5c3f75c9ee0bb6da69ff38e4f7de4bc9b8b632fd 
```

You should see an editor pop up with these contents:

```
pick d03596eb "Add / Hide Tokens" screen "Add /...
pick 0c06f1e2 code refactor, bags fix
pick 5e54ab6c remove unused pod, fixed bugs, refactored code
pick 75ae1d15 Minor refactor and bugs fix: - rena...
```

Note that this is newest commit at the bottom.

Modify this file so that those 4 lines above become:

```
pick d03596eb "Add / Hide Tokens" screen "Add /...
squash 0c06f1e2 code refactor, bags fix
squash 5e54ab6c remove unused pod, fixed bugs, refactored code
squash 75ae1d15 Minor refactor and bugs fix: - rena...
```

(You can ignore the lines that start with #, git igores those) and save the file. git should continue and show another text file to edit, starting with something like:

```
# This is a combination of 4 commits.                                                                                                                                                      
# This is the 1st commit message:

"Add / Hide Tokens" screen
"Add / Hide Tokens" button in the Wallet tab
"Hide" button, and add hide token logic in the Wallet tab
Implement search on "Add / Hide Tokens" screen
Move add token logic to "Add / Hide Tokens" screen

 # This is the commit message #2:
 
 code refactor, bags fix
 
 # This is the commit message #3:
 
 remove unused pod, fixed bugs, refactored code
 
 It looks like TokensCoordinator.didDelete() isn't used anymore.
```

Remember that lines starting with # are ignored, so remove all the lines in that file (you can keep the lines with # if you like, they get ignored) and replace it with just 1 line (no line break):

```
Add "Add / Hide Tokens" screen for token management
```

Save and git should proceed to rebase.

Run this again:

```
git log --pretty=oneline add_hide_tokes
```

and you should see (with a different hash)

```
bb83a148ca2db7002226463694894aafef135411 (HEAD -> add_hide_tokes) Add "Add / Hide Tokens" screen for token management
5c3f75c9ee0bb6da69ff38e4f7de4bc9b8b632fd Fixed src link of main image
```


Notice the commit 1st commit bb83a148ca2db7002226463694894aafef135411 (you will have a different hash). So the 4 commits have been squashed into 1. Now force push (make sure you replace `origin` with the correct remote that you normally push to):

```
git push origin add_hide_tokes --force
```

vladyslav-iosdev commented:
@hboon 
about this "But before I forget, we talked about combining the database migration versions into 1, instead of 6, 7, 8 (I think?), can you just keep them as 6?"
i tried to left 6 instead of 8, but in this case i get crash on migration, that is why i left it as 6.
I forgot mentioned about it

hboon commented:
> @hboon
> about this "But before I forget, we talked about combining the database migration versions into 1, instead of 6, 7, 8 (I think?), can you just keep them as 6?"
> i tried to left 6 instead of 8, but in this case i get crash on migration, that is why i left it as 6.
> I forgot mentioned about it

@vladyslav-iosdev Thanks for checking back. Yes, I mean to combine 6, 7, 8. We'll have to delete the app on devices which installed previous builds with this migration version 6, 7, 8 in order to work. That's usually ok, because the code isn't public yet.

I'll take over from now and finish up the squashing and do a final pass. But otherwise it looks good already. Thanks for the good work. 

I'll likely create a new PR with your commits so I can push to the branch and @ you. Feel free to watch out for it for completion sake, but you don't need to do anything.

vladyslav-iosdev commented:
> Thanks for checking back. Yes, I mean to combine 6, 7, 8. We'll have to delete the app on devices which installed previous builds with this migration version 6, 7, 8 in order to work. That's usually ok, because the code isn't public yet.
> 
> I'll take over from now and finish up the squashing and do a final pass. But otherwise it looks good already. Thanks for the good work.
> 
> I'll likely create a new PR with your commits so I can push to the branch and @ you. Feel free to watch out for it for completion sake, but you don't need to do anything.

Hey, changes uploaded.

hboon commented:
@vladyslav-iosdev thanks!

hboon commented:
@James-Sangalli can you help to review?

A few things to look out for:

1\. There's a database migration
2\. Verify that upon installation of a build with this PR, tokens that already exist should still be visible and in the same order as pre-PR. One way to check is to go into Add / Hide Tokens screen and scroll to the bottom. There should no tokens under "Hidden Tokens".

James-Sangalli commented:
> @James-Sangalli can you help to review?
> 
> A few things to look out for:
> 
> 1. There's a database migration
> 2. Verify that upon installation of a build with this PR, tokens that already exist should still be visible and in the same order as pre-PR. One way to check is to go into Add / Hide Tokens screen and scroll to the bottom. There should no tokens under "Hidden Tokens".

Was able to install without issue. number 2 is also fine.
